### PR TITLE
Implements _Launch on App Engine_

### PR DESCRIPTION
**Pull request by briandealwis**
- Creates a new command `com.google.cloud.tools.eclipse.appengine.localserver.launch` that takes a parameter as to whether it should run or debug.  This command can be invoked from the Quick Access.
  - Provides a single handler that supports either the selected projects (e.g., from the _Project Explorer_) or if the project containing the file in the current editor.
  - extracts a few helper methods from elsewhere to aid in testing

Part of #225

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    IFile descriptorFile = webinf.getFile("appengine-web.xml");
+    assertTrue(descriptorFile.exists());
+    descriptorFile.setContents(appEngineForService(serviceId), IFile.FORCE, null);
+
+    return module;
+  }
+
+  private static InputStream appEngineForService(String serviceId) {
+    StringBuilder contents = new StringBuilder();
+    contents.append("<?xml version=\"1.0\" encoding=\"utf-8\"?>\n");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java]: This line isn't needed here, and the test is easier to read without it.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    contents.append("<appengine-web-app xmlns=\"http://appengine.google.com/ns/1.0\">\n");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java]: You can use single quotes around the attribute values to make the embedded strings more legible.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    contents.append("<service>").append(serviceId).append("</service>\n");
+    contents.append("</appengine-web-app>\n");
+    return new ByteArrayInputStream(contents.toString().getBytes(StandardCharsets.UTF_8));
+  }
+
+  @Test
+  public void testWithOneModule() throws ExecutionException, CoreException {
+    module1 = mockAppEngineStandardModule("default", appEngineStandardProject1.getProject());
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java]: If you're always initializing module1 and module2 in test methods, you might as well make them local variables.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    ExecutionEvent event =
+        new ExecutionEventBuilder().withCurrentSelection(module1, module2).build();
+    handler.execute(event);
+    assertEquals("new server should have been created", 1, tracker.getServers().size());
+
+    handler.execute(event);
+    assertEquals("no new server should be created", 1, tracker.getServers().size());
+
+    event = new ExecutionEventBuilder().withCurrentSelection(module2, module1).build();
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java]: safer not to reuse local variables.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerTracker.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.List;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.wst.server.core.IServer;
+import org.eclipse.wst.server.core.IServerLifecycleListener;
+import org.eclipse.wst.server.core.ServerCore;
+import org.junit.rules.ExternalResource;
+
+/** Track creation of WTP {@link IServer} instances and ensure they are deleted. */
+public class ServerTracker extends ExternalResource {
+  private List<IServer> servers = new ArrayList<>();
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerTracker.java]: I wouldn't be surprised if this has to be thread safe.

**Comment by briandealwis** [comment]
Turns out there's little point mocking up `IModule` when we've created a project on-disk.  Also saw a failure on Neon which is likely due to mock comparisons.

File: plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ui/ExecutionEventBuilder.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    return new ExecutionEvent(null /* command */, Collections.EMPTY_MAP, null /* trigger */,
+        context);
+  }
+
+  public ExecutionEventBuilder withActiveShell(Shell shell) {
+    when(context.getVariable(ISources.ACTIVE_SHELL_NAME)).thenReturn(shell);
+    return this;
+  }
+
+  public ExecutionEventBuilder withCurrentSelection(Object object) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ui/ExecutionEventBuilder.java]: There a reason for a single-arg version considering that there is a multi-arg one below?

File: plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/project/TestProjectCreator.java [hunk 7]
```diff
@@ -52,6 +62,11 @@ public TestProjectCreator withFacetVersions(List<IProjectFacetVersion> projectFa
     return this;
   }
 
+  public TestProjectCreator withAppEngineServiceId(String serviceId) {
+    this.appEngineServiceId = serviceId;
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/project/TestProjectCreator.java]: `this.` necessary?

Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/project/TestProjectCreator.java]: It can go.

**Comment by briandealwis** [comment]
PTAL @elharo @chanseokoh

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
+
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java]: copyright please

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    return null;
+  }
+
+  private IServer createServer(IModule[] modules, IProgressMonitor monitor) throws CoreException {
+    SubMonitor progress = SubMonitor.convert(monitor, 10);
+    IServerType serverType = ServerCore.findServerType(LocalAppEngineServerDelegate.SERVER_TYPE_ID);
+    IServerWorkingCopy serverWorkingCopy =
+        serverType.createServer(null, null, progress.newChild(4));
+    // serverWorkingCopy.setAttribute("name", "Generated server instance");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java]: delete

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  public TestProjectCreator appEngineStandardProject1 =
+      new TestProjectCreator().withFacetVersions(Lists.newArrayList(JavaFacet.VERSION_1_7,
+          WebFacetUtils.WEB_25, APPENGINE_STANDARD_FACET_VERSION_1));
+  @Rule
+  public TestProjectCreator appEngineStandardProject2 =
+      new TestProjectCreator().withFacetVersions(Lists.newArrayList(JavaFacet.VERSION_1_7,
+          WebFacetUtils.WEB_25, APPENGINE_STANDARD_FACET_VERSION_1));
+
+  @Before
+  public void setUp() throws CoreException {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java]: `throws CoreException` unnecessary.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerDelegate.java [hunk 11]
```diff
@@ -41,6 +41,11 @@
 
 @SuppressWarnings("restriction") // For FacetUtil
 public class LocalAppEngineServerDelegate extends ServerDelegate {
+  public static final String RUNTIME_TYPE_ID =
+      "com.google.cloud.tools.eclipse.appengine.standard.runtime";
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerDelegate.java]: FYI, this is also defined in `AppEngineStandardFacet.java`.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+  @Override
+  public Object execute(ExecutionEvent event) throws ExecutionException {
+    String launchMode = event.getParameter("launchMode");
+    if (launchMode == null) {
+      launchMode = ILaunchManager.DEBUG_MODE;
+    }
+    IModule[] modules = asModules(event);
+    SubMonitor progress = SubMonitor.convert(null, 10);
+    try {
+      IServer server = findExistingServer(modules, progress.newChild(3));
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java]: What if we found an existing server, which is running in RUN_MODE, when a user wanted to run in DEBUG_MODE? In this case, I think this code does nothing; there won't be a server running in DEBUG_MODE.

Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java]: The mode doesn't actually matter when finding a server.

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java]: Yeah, that's what I'm talking. I may be missing something, but if we are in the case I described above, shouldn't we either stop the existing server and restart it in the right mode, or create a new server in the right mode?

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerDelegate.java (hunk 11 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerDelegate.java]: Oh geesh, this is quite circular.  `.localserver` imports classes from `.facets`, but `.facets` uses runtime and server definitions from `.localserver`.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java (hunk 12 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java]: The Server infra doesn't allow multiple server launches.  I don't think we want to recreate the server.  But offering to relaunch the server sounds like a good idea. At least alerting the user to that fact.

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandler.java]: Yeah, if it's difficult to relaunch the server (for now), the simplest solution could just be showing an alert that you already have a running server in a different mode and should stop it first.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+      new TestProjectCreator().withFacetVersions(Lists.newArrayList(JavaFacet.VERSION_1_7,
+          WebFacetUtils.WEB_25, APPENGINE_STANDARD_FACET_VERSION_1));
+  @Rule
+  public TestProjectCreator appEngineStandardProject2 =
+      new TestProjectCreator().withFacetVersions(Lists.newArrayList(JavaFacet.VERSION_1_7,
+          WebFacetUtils.WEB_25, APPENGINE_STANDARD_FACET_VERSION_1));
+
+
+  @Before
+  public void setUp() throws CoreException {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java]: `throws CoreException` unnecessary.

**Comment by briandealwis** [comment]
Meant to show the output when a server is already launched.

If in same mode:
screen shot 2016-12-08 at 2 22 51 pm

In different launch mode:
screen shot 2016-12-08 at 2 22 26 pm

**Comment by briandealwis** [comment]
@PTAL @elharo

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+      protected IServer findExistingServer(IModule[] modules, SubMonitor progress) {
+        if (serverToReturn != null) {
+          return serverToReturn;
+        }
+        return super.findExistingServer(modules, progress);
+      }
+
+    };
+  }
+
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java]: nit: extra blank line

**Pull request merged**